providers/component-protocol/definition/scenario_render.go

```diff
@@ -0,0 +1,386 @@
+// Copyright (c) 2021 Terminus, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package definition
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/erda-project/erda-infra/providers/component-protocol/definition/cptype"
+)
+
+// CompRenderSpec .
+type CompRenderSpec struct {
+	// 具体的场景名
+	Scenario string `json:"scenario"`
+	// 具体的组件名
+	CompName string `json:"name"`
+	// 组件（包含渲染函数）创建函数
+	RenderC RenderCreator
+}
+
+// RenderCreator .
+type RenderCreator func() CompRender
+
+// CompRender .
+type CompRender interface {
+	Render(ctx context.Context, c *cptype.Component, scenario cptype.ComponentProtocolScenario, event cptype.ComponentEvent, gs *cptype.GlobalStateData) error
+}
+
+// ScenarioRender .
+type ScenarioRender map[string]*CompRenderSpec
+
+// ScenarioRenders .
+var ScenarioRenders = make(map[string]*ScenarioRender)
+
+// MustRegister .
+func MustRegister(r *CompRenderSpec) {
+	if err := Register(r); err != nil {
+		panic(err)
+	}
+}
+
+// Register .
+func Register(r *CompRenderSpec) error {
+	if r == nil {
+		err := fmt.Errorf("invalid register request")
+		return err
+	}
+	if r.Scenario == "" || r.CompName == "" {
+		err := fmt.Errorf("invalid register request, empty scenario [%s] or compName [%s]", r.Scenario, r.CompName)
+		return err
+	}
+
+	logrus.Infof("start to register scenario:%s", r.Scenario)
+	// if scenario key not exit, crate it
+	if _, ok := ScenarioRenders[r.Scenario]; !ok {
+		s := make(ScenarioRender)
+		ScenarioRenders[r.Scenario] = &s
+	}
+	// if compName key not exist, create it and the CompRenderSpec
+	s := ScenarioRenders[r.Scenario]
+	if _, ok := (*s)[r.CompName]; !ok {
+		(*s)[r.CompName] = r
+	} else {
+		err := fmt.Errorf("register render failed, component [%s] already exist", r.CompName)
+		return err
+	}
```

Comment:
component should be registered in a bigger namespace, not just place in one scenario.